pkg/services/dummy/dummy.go

```diff
@@ -0,0 +1,215 @@
+	return s.sessionID
+}
+
+func (s sessionPointer) Payload() []byte {
+	return []byte("dummy")
+}
+
+func (s sessionPointer) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		SessionID string `json:"sessionID"`
+	}{SessionID: s.sessionID})
+}
+
+type signingSessionResult struct {
+	ID      string
+	State   string
+	Request string
+}
+
+func (d signingSessionResult) Status() string {
+	return d.State
+}
+
+func (d signingSessionResult) VerifiablePresentation() (contract.VerifiablePresentation, error) {
+	// todo: the contract template should be used to select the dummy attributes to add
+	// reqContract := d.Request
+
+	if d.Status() != SessionCompleted {
+		return nil, nil
+	}
+
+	return Presentation{
+		VerifiableCredentialBase: contract.VerifiableCredentialBase{
+			Context: []string{contract.VerifiableCredentialContext},
+			Type:    []string{contract.VerifiablePresentationType, VerifiablePresentationType},
+		},
+		Proof: Proof{
+			Type:      NoSignatureType,
+			Initials:  "I",
+			Lastname:  "Tester",
+			Birthdate: "1980-01-01",
+			Email:     "tester@example.com",
+			Contract:  d.Request,
+		},
+	}, nil
+}
+
+var errNotEnabled = errors.New("not allowed in strict mode")
+
+// VerifyVP verifies a verifiablePresentation provided by a byte array.
+// It returns the contract.VerificationResult.
+func (d Dummy) VerifyVP(rawVerifiablePresentation []byte) (*contract.VerificationResult, error) {
+	if d.InStrictMode {
+		return nil, errNotEnabled
+	}
+
+	ndp := Presentation{}
+	if err := json.Unmarshal(rawVerifiablePresentation, &ndp); err != nil {
+		return nil, err
+	}
+
+	c, err := contract.ParseContractString(ndp.Proof.Contract, contract.StandardContractTemplates)
+	if err != nil {
+		return nil, err
+	}
+
+	return &contract.VerificationResult{
+		State:          contract.Valid,
+		ContractFormat: ContractFormat,
+		DisclosedAttributes: map[string]string{
+			"initials":  ndp.Proof.Initials,
+			"lastname":  ndp.Proof.Lastname,
+			"birthdate": ndp.Proof.Birthdate,
+			"email":     ndp.Proof.Initials,
+		},
+		ContractAttributes: c.Params,
+	}, nil
+}
+
+// SigningSessionStatus looks up the session by the provided sessionID param.
+// When the session exists it returns the current state and advances the state to the next one.
+// When the session is SessionComplete, it removes the session from the sessionStore.
+func (d Dummy) SigningSessionStatus(sessionID string) (contract.SigningSessionResult, error) {
+	if d.InStrictMode {
+		return nil, errNotEnabled
+	}
+
+	state, ok := d.Status[sessionID]
+	if !ok {
+		return nil, services.ErrSessionNotFound
+	}
+
+	session, ok := d.Sessions[sessionID]
+	if !ok {
+		return nil, services.ErrSessionNotFound
+	}
+
+	// increase session status everytime this request is made
+	switch state {
+	case SessionCreated:
+		d.Status[sessionID] = SessionInProgress
+	case SessionInProgress:
+		d.Status[sessionID] = SessionCompleted
+	case SessionCompleted:
+		delete(d.Status, sessionID)
```

Comment:
> SigningSessionStatus looks up the session by the provided sessionID param.

(function proceeds to modify its state)

If a function's documentation tells me it looks something up, I don't expect it to modify its state. So either:

- The function's name is wrong
- The implementation is wrong
- The documentation is wrong

(or a combination)

Reply:
The dummy only simulates a Signer. It can't be used in production. Part of its simulation is to go through different states. The func name cannot be changed since it's an interface.